compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -826,9 +835,27 @@ checkTemplate m t@(Template _loc tpl param precond signatories observers text ch
     withPart TPAgreement $ checkExpr text TText
     for_ choices $ \c -> withPart (TPChoice c) $ checkTemplateChoice tcon c
   whenJust mbKey $ checkTemplateKey param tcon
+  forM_ implements $ checkIfaceImplementation (moduleName m) tcon
   where
     withPart p = withContext (ContextTemplate m t p)
 
+checkIfaceImplementation ::
+     MonadGamma m
+  => ModuleName
+  -> Qualified TypeConName
+  -> Qualified TypeConName
+  -> m ()
+checkIfaceImplementation modName tplTcon ifTcon
+  | Qualified PRSelf m _tconName <- ifTcon
+  , m == modName = do
+    DefInterface {intChoices} <- inWorld $ lookupInterface ifTcon
+    forM_ intChoices $ \InterfaceChoice {ifcName, ifcConsuming, ifcArgType, ifcRetType} -> do
+      TemplateChoice {chcConsuming, chcArgBinder, chcReturnType} <- inWorld $ lookupChoice (tplTcon, ifcName)
+      unless (chcConsuming == ifcConsuming) $ throwWithContext $ EBadInterfaceChoiceImplConsuming ifcName ifcConsuming chcConsuming
+      unless (snd chcArgBinder == ifcArgType) $ throwWithContext $ EBadInterfaceChoiceImplArgType ifcName ifcArgType (snd chcArgBinder)
```

Comment:
We should be using `alphaType` here instead of structural equality. Doesn't matter too much for now but  better to be on the safe side.

Reply:
agreed, changed it.